internal/controllers/core/kubernetesdiscovery/portforwards.go

```diff
@@ -0,0 +1,180 @@
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			errs = append(errs, fmt.Errorf("creating portforward %s: %v", pf.Name, err))
+		}
+	}
+
+	return errorutil.NewAggregate(errs)
+}
+
+// Construct the desired port-forward. May be nil.
+func (r *Reconciler) toDesiredPortForward(kd *v1alpha1.KubernetesDiscovery) (*v1alpha1.PortForward, error) {
+	pfTemplate := kd.Spec.PortForwardTemplateSpec
+	if pfTemplate == nil {
+		return nil, nil
+	}
+
+	pod := pickBestPortForwardPod(kd)
+	if pod == nil {
+		return nil, nil
+	}
+
+	pf := &v1alpha1.PortForward{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: pod.Name,
+			Annotations: map[string]string{
+				v1alpha1.AnnotationManifest: kd.Annotations[v1alpha1.AnnotationManifest],
+				v1alpha1.AnnotationSpanID:   kd.Annotations[v1alpha1.AnnotationSpanID],
+			},
+		},
+		Spec: v1alpha1.PortForwardSpec{
+			PodName:   pod.Name,
+			Namespace: pod.Namespace,
+			Forwards:  pfTemplate.Forwards,
+		},
+	}
+	populateContainerPorts(pf, pod)
+	err := controllerutil.SetControllerReference(kd, pf, r.ctrlClient.Scheme())
+	if err != nil {
+		return nil, err
+	}
+	return pf, nil
+}
+
+// If any of the port-forward specs have ContainerPort = 0, populate them with
+// the documented ports on the pod. If there's no default documented ports for
+// the pod, populate it with the local port.
+//
+// TODO(nick): This is old legacy behavior, and I'm not totally sure it even
+// makes sense. I wonder if we should just insist that ContainerPort is populated.
+func populateContainerPorts(pf *v1alpha1.PortForward, pod *v1alpha1.Pod) {
+	cPorts := store.AllPodContainerPorts(*pod)
+	for i, forward := range pf.Spec.Forwards {
+		if forward.ContainerPort == 0 && len(cPorts) > 0 {
+			forward.ContainerPort = int32(cPorts[0])
+			for _, cPort := range cPorts {
+				if int(forward.LocalPort) == int(cPort) {
+					forward.ContainerPort = int32(cPort)
+					break
+				}
+			}
+		}
+		if forward.ContainerPort == 0 {
+			forward.ContainerPort = forward.LocalPort
+		}
+		pf.Spec.Forwards[i] = forward
+	}
+}
+
+// We can only portforward to one pod at a time.
+// So pick the "best" pod to portforward to.
+// May be nil if there is no eligible pod.
+func pickBestPortForwardPod(kd *v1alpha1.KubernetesDiscovery) *v1alpha1.Pod {
+	var bestPod *v1alpha1.Pod
+	for _, pod := range kd.Status.Pods {
+		pod := pod
+		if pod.Name == "" {
+			continue
+		}
+
+		// Only do port-forwarding if the pod is running or deleting.
+		isRunning := pod.Phase != string(v1.PodRunning)
+		isDeleting := pod.Deleting
+		if !isRunning && !isDeleting {
+			continue
+		}
+
+		// This pod is eligible! Now compare it to the existing candidate (if there is one).
+		if bestPod == nil || isBetterPortForwardPod(&pod, bestPod) {
+			bestPod = &pod
+		}
+	}
+	return bestPod
+}
+
+// Check if podA is better than podB for port-forwarding.
+func isBetterPortForwardPod(podA, podB *v1alpha1.Pod) bool {
+	// A deleting pod is always better than a non-deleting pod.
```

Comment:
```suggestion
	// A non-deleting pod is always better than a deleting pod.
```
I might have not yet had enough coffee, but I think this should be the opposite? That seems to be what the logic is actually doing - comment just caught me off-guard 😛

Reply:
oops, fixed